### ISSUE TITLE
[NuGet.org Bug]: NuGet.org tampers with the uploaded package, making attestation difficult

### ISSUE THREAD

**Issue by Smaug123** [NuGet/NuGetGallery#10026]
### Impact

It bothers me. A fix would be nice

### Describe the bug

I recently implemented [GitHub Attestations](https://github.blog/2024-05-02-introducing-artifact-attestations-now-in-public-beta/) for the NuGet packages I produce. This was pretty annoying to do, because NuGet.org inserts a signature file into any(?) nupkg you upload to it. So the only flow I can make work is:

* upload the package;
* wait for it to become available;
* download the package; 
* unzip it and verify that its contents are byte-for-byte identical after removing the .signature file;
* attest the package I downloaded.

This is awkward!

### Repro Steps

Upload a package to NuGet. Download it again. Observe that its contents have changed.

### Expected Behavior

It surprised me greatly that NuGet.org tampers with uploaded packages. (The obvious thing to do would have been to have NuGet serve a certificate separately from the package, but that ship has presumably long since sailed. "NuGet.org attests that this package was uploaded by someone capable of impersonating ThisParticularUser" is not information that belongs *in* the package; other package repositories do exist and it would be nice not to require each one to tamper incompatibly with the packages that get uploaded to it to attest to it!)

No solution comes to mind; it may be possible for NuGet.org *not* to tamper with uploaded packages, but I can't see how without either duplicating large swathes of API or breaking compatibility.

### Additional Context and logs

It is of course possible to sign NuGet packages, but this has exactly the same secrets-management problems that NuGet API keys have. The whole point is to attest meaningfully that _this package_ was produced from _this source_; having to trust that package producers are correctly handling their certificates is a very big and unnecessary ask. (Of course, better would be to have reproducible NuGet packs, because then we could attest verifiably that a package was built from a particular source tree *without* a bunch of custom logic.)

**Comment by Smaug123** [comment]
(This isn't *really* a bug, but I didn't think any of the labels were appropriate. "Design decisions are making things much harder than they needed to be".)

**Comment by Smaug123** [comment]
My purpose in attestation here is to credibly claim that an open-source package has the contents it appears to have. I can claim that package Foo.Bar was built from Git source abc123 without modification, but (as far as I know) the NuGet ecosystem provides no way for me to make this claim *credibly*. It is possible in principle for me to use the existing NuGet signature mechanism to assert that I produced the package contents, but as far as I can tell there is no existing way for me to assert that the package contents are what I claim they are.

However, with GitHub's artifact attestations, I *can* credibly claim that _this_ package was produced by _this_ automated pipeline run from _this_ git source tree which anyone can audit. GitHub's artifact attestations are of the form "an artefact with this SHA256 has this provenance", and as far as I can tell it does not admit special-casing for NuGet; it would take bespoke tooling to phrase "these packages are equal *modulo bundled signature files*" in this form.

**Comment by Smaug123** [comment]
To be extremely concrete, I have some packages that I wrote myself which I wish to use in a context where the integrity of the code we're running is important: it's unacceptable for us to run packages for whose contents we have only my word. We *could* clone the packages from source, run parallel publishing pipelines in contexts that we trust, and so on; or we could remove the need to trust me, and instead only trust GitHub (which is much simpler).

**Comment by joelverhagen** [comment]
Thanks for the additional information.

In the NuGet signing "parallel universe" this credible claim you mention is analogous to NuGet author signatures. If you, the package author, use `dotnet nuget sign` I can credibly claim that the package was signed by the owner of that author signing certificate. To avoid trusting you, your organization (presumably the "we" in your last message) could entrust a code signing certificate to GitHub Actions (however you choose) and author sign the package. The package will be modified on NuGet.org but you can still verify the original author signature is valid. Of course you need to trust NuGet client tooling in this case.

A workaround for this difference in approach could be to simply not publish the NuGet package to a repository which supports repository signing (NuGet.org is the only I know of which does). For your concrete case, this could take the form of publishing the packages once to NuGet.org (if desired) for external/OSS consumption. Then you would publish the original packages to another feed, perhaps Azure Artifacts or GitHub Package Registry, and consuming from there for your highly sensitive context.

What's not clear to me is how you are performing the "verify" step of the consumed artifacts. Are you using both NuGet restore from NuGet.org and additionally `gh attestation verify` (per [docs](https://docs.github.com/en/actions/security-guides/using-artifact-attestations-to-establish-provenance-for-builds#verifying-artifact-attestations-with-the-github-cli))?

Where is the verify happening in your consumption workflow? I assume you are using `actions/attest-build-provenance@v1` against the .nupkg after `dotnet pack` in your GitHub Actions workflow.

**Comment by joelverhagen** [comment]
If you are populating and then consuming from an intermediate cache, then I imagine the source to populate the cache is less important.

If you don't consume directly from your cache, there is no guarantee the content fetched from the original source has the same attestation state at the time of package restore (`dotnet restore`). Therefore you'll need to set your NuGet package source as the cache.

This leaves just the matter of populating the cache, in which case NuGet.org doesn't seem to be unique. You just need some tooling (perhaps NuGet itself with an overridden `NUGET_PACKAGES` environment variable) to download your dependency graph from somewhere. Then the attest step. Then the `nuget restore` from your cache dir.

It seems like there are some open questions on how the end-to-end will work. It appears to me that non-trivial custom tooling is needed as well as new mappings (e.g. associating repository X with .nupkg Y). It's also not clear how transitive dependencies that you don't own will be attested.

I'll leave this issue open since it is indeed true, attestation based on trivial file hash is indeed hard to do with repository-signed NuGet packages.

**Comment by Smaug123** [comment]
Yeah, the obvious proof-of-concept for this is [the standard .NET workflow on NixOS](https://github.com/NixOS/nixpkgs/blob/6efefdc8c16fe1c483f35e3c5f6c1de45f5f4e23/pkgs/build-support/dotnet/build-dotnet-module/default.nix), which fetches every package from [a lockfile](https://github.com/Smaug123/WoofWare.Myriad/blob/669eccbdef483a904ebed8b186208203925879d6/nix/deps.nix) into an isolated environment and [sets the NuGet source to that environmnet](https://github.com/NixOS/nixpkgs/blob/6efefdc8c16fe1c483f35e3c5f6c1de45f5f4e23/pkgs/build-support/dotnet/build-dotnet-module/hooks/dotnet-configure-hook.sh#L37) before build. This also asserts the contents of the package as a side-effect.

Transitive dependencies that I don't own are certainly important, and it would be great to see an attestation solution across the ecosystem! Some dependencies we're happy to accept on faith - e.g. FSharp.Core is already author-signed, and we are happy to assume its author is taking care of the certificate - but most dependencies are currently only dubiously acceptable or (e.g. in my case, an insider who is the sole author of an opaque binary blob) are completely unacceptable. That's why I was so excited to learn about attestations: it finally gives me the technology to prove that my binary blob corresponds to a given reviewable source, which makes it in principle acceptable to consume.

If I had god perms to the universe, I'd allow NuGet.org to associate packages with repos. Packages already can declare their source, and NuGet.org provides links to the source; it would (for example) be trivial for NuGet to check with GitHub at upload time that the uploaded package has been attested at the version it claims to have. (But for this feature to be useful enough to implement, it would require massive vendor lock-in to GitHub - I'm not aware of other ecosystems which allow such trivial attestation - and would require "most" package authors to have unusually secure/trustworthy publishing workflows.)

**Comment by Smaug123** [comment]
The instability of NuGet package SHAs, by the way, is aesthetically displeasing for Nix as well. It *should*, morally, be the case that I can build a package from source and have the result be equal to the package I obtain from NuGet; if that were true and if `dotnet pack` were reproducible, I could build the package into my Nix store from source and then immediately consume it without waiting for a round-trip to NuGet.org to find a SHA to update.

**Comment by joelverhagen** [comment]
Even without package signing, I don't believe building .NET packages from source is byte-for-byte reproducible, but this is not an area I have investigated very much. I found https://github.com/dotnet/reproducible-builds via a Google search so perhaps you can give that a try. In general, I think byte-for-byte (i.e. hash equivalent) building of a package from source is not something that many customers have asked for so I don't have the experience of trying it and seeing what works and what doesn't.

**Comment by erdembayar** [comment]
> Even without package signing, I don't believe building .NET packages from source is byte-for-byte reproducible, but this is not an area I have investigated very much. 

In my opinion, this byte-for-byte reproducible approach may work in the short term, but most likely will not work after a few months. Most NuGet package dependencies are version ranges, so the package dependencies you have today may change at any time when older dependencies are unlisted and a new patched version is published. Of course, you can set a specific version and use a lock file to ensure you're using the same upstream package, but it requires quite an effort to make it happen for an end-to-end scenario. Many dotnet repositories that are a few years old are not rebuildable at all nowadays from source code.

**Comment by Smaug123** [comment]
Yeah, NixOS's setup gives that stuff for free. I guarantee the "isolated" property of the Dotnet.ReproducibleBuilds.Isolated package by packaging with Nix, and I lock the SDK and all NuGet packages by hash at the Nix level for the build; it works pretty seamlessly. `ContinuousIntegrationBuild=true` and `Deterministic=true` appear in my experience to be enough to keep the output stable under those conditions, so it's only the timestamps in the .nupkg file that are nondeterministic.

**Comment by Smaug123** [comment]
Is there anywhere I can see NuGet.org's logic for adding this signature file? Empirically I have observed it to be the case that `zip -d "$downloaded_package" .signature.p7s` recovers precisely the original file, but there doesn't seem to be any particular reason to expect that *would* be true; in principle the whole thing could be recompressed when the `.signature.p7s` is added. I've completely failed to find the relevant logic in NuGetGallery. (It would be *very* helpful if we could reliably obtain the original package, given the one NuGet serves us.)

**Comment by joelverhagen** [comment]
> Empirically I have observed it to be the case that `zip -d "$downloaded_package" .signature.p7s` recovers precisely the original file, but there doesn't seem to be any particular reason to expect that *would* be true

This can be true. `zip -d` would need to implement file removal in a specific way, as to not disturb the other file entries in a way that changes the NuGet content hash. If you are seeing that behavior, it is a fortunate coincidence.

For packages that have an existing `.signature.p7s` file, this would not work. This most often occurs when the package is author signed. In this case the `.signature.p7s` file is modified to contain a repository counter signature.

> in principle the whole thing could be recompressed when the .signature.p7s is added. 

This would be a breaking change because the stable "content hash" that is included in the signature (and is part of the NuGet package signing contract) would change. 

> I've completely failed to find the relevant logic in NuGetGallery. 

NuGetGallery open source code will not show this part. We have a small closed source integration with Microsoft internal signing systems which provides repository signing. This closed source integration does internally use the code you linked (via [`MSSignCommand`](https://github.com/NuGet/NuGet.Client/blob/128a5066b1438627ac69a2ffe9de564b2c09ee4d/src/NuGet.Clients/NuGet.MSSigning.Extensions/MSSignCommand.cs) I think), but there are a couple of layers of abstraction between NuGet.org and the code link you found.

> It would be very helpful if we could reliably obtain the original package, given the one NuGet serves us.

This is an interesting request and, if you want it, should probably be its own issue. Feel free to open this other issue and cross link it to this one! If I remember correctly, we do store the original artifact before repository signing so we could theoretically make this available for both existing and new packages. But there may be caveats that I don't know of right now, not found until such a feature request is designed and implemented.

**Comment by Smaug123** [comment]
Thanks, that comment clarifies something that in hindsight was fuzzy in my mind. I care about *determinism*, not about retrieving the original package per se; so stripping all signatures is fine for my purposes.

I dug through `zip` and, as long as the original package was unsigned, `zip -d` does seem to be [simply inverting what NuGet does](https://github.com/darlinghq/darling-zip/blob/caf41ebbc3ebab0250e4d13aa42221ef91a9802c/zip/zip30/zip.c#L4268) (although I can't say I fully understand that code). I think for nixpkgs we can work with what we have! (But only because in nixpkgs we get the chance to intervene before any consumer sees the nupkg; that's not really true in other contexts.)

**Comment by Frulfump** [comment]
Not sure if this is relevant here but could be https://github.com/NuGet/Home/issues/14411

**Comment by aetos382** [comment]
Just a thought, instead of automatically signing packages when pushing to `nuget.org`, how about providing a feature like `nuget sign -remote`?
This would call the `nuget.org` service from within GitHub Actions, retrieve the signed package, and then generate the attestation.

**Comment by Smaug123** [comment]
I guess that would make my current workflow faster because I currently have to wait for the package to become available for download from nuget.org; presumably `dotnet nuget sign` would e.g. allow me to skip indexing in nuget.org's backend (as long as the `dotnet nuget sign`'ed package is binary-identical when I come to `dotnet nuget push` it later). It doesn't relieve me of the need to check that the contents of the signed package are the same as what I uploaded.

**Comment by sgbench** [comment]
I share the concerns of the others here, and if the current signing behavior of nuget.org is truly set in stone, then something along the lines of @aetos382's suggestion (being able to seamlessly attest the nuget.org-signed .nupkg from within CI) is probably the best approach. Being able to download the original .nupkg from nuget.org would be an improvement over the current situation, yes, but the consumer would still need to either:

1. Define a local NuGet package source containing these unsigned .nupkg files, in which case the consuming tools don't benefit from the signatures added by nuget.org, or
2. Also download the signed versions and verify that the contents (minus the signature file) match before adding the signed .nupkg files to a local NuGet package source, which is a lot of work that shouldn't be necessary.

If the .NET ecosystem (starting with NuGet) doesn't take steps to reduce the friction of supporting build attestations, or takes too little action too late, then the practice will go largely unadopted and the ecosystem will suffer as a result. Supply chain is the new weak link for many high-value targets, and I would love to see .NET position itself ahead of the curve here.

@Smaug123 Another potential workaround would be for the CI workflow to pack the project, unzip the resulting .nupkg, attest all the extracted files (in a single attestation), then push to nuget.org. Consumers would download the signed .nupkg from nuget.org and verify all extracted files (minus the signature file). Though I suppose this would mostly make things harder for consumers and wouldn't simplify the CI workflow very much.

**Comment by Smaug123** [comment]
Yeah, I much prefer a path which doesn't require my consumers to write custom tooling specifically for the NuGet ecosystem.
